### PR TITLE
CRM-19827 Avoid unnecessary core_resources processing during ajax calls

### PR DESCRIPTION
**Pull request by colemanw**
* [CRM-19827: Avoid unnecessary core_resources processing during ajax calls](https://issues.civicrm.org/jira/browse/CRM-19827)

### CODE REVIEW

File: CRM/Utils/System.php [hunk 1]
```diff
@@ -301,6 +301,16 @@ public static function url(
...
 
   /**
+   * @return string|null
+   */
+  public static function getUrlPath() {
+    if (isset($_GET[CRM_Core_Config::singleton()->userFrameworkURLVar])) {
+      return $_GET[CRM_Core_Config::singleton()->userFrameworkURLVar];
+    }
+    return NULL;
+  }
```
Comment (colemanw) [inline CRM/Utils/System.php]: Can't believe we didn't already have this utility.

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Utils/System.php (hunk 1 above)
Comment (eileenmcnaughton) [inline CRM/Utils/System.php]: So in the spirit of our recent comments fix ups the comment should have a sentence description to meet the full drupal standard

File: CRM/Core/Controller.php [hunk 2]
```diff
@@ -457,8 +457,8 @@ public function addPages(&$stateMachine, $action = CRM_Core_Action::NONE) {
       if ($options) {
         $$stateName->setOptions($options);
       }
-      if (property_exists($$stateName, 'urlPath') && isset($_GET[CRM_Core_Config::singleton()->userFrameworkURLVar])) {
-        $$stateName->urlPath = explode('/', $_GET[CRM_Core_Config::singleton()->userFrameworkURLVar]);
+      if (property_exists($$stateName, 'urlPath')) {
```
Comment (eileenmcnaughton) [inline CRM/Core/Controller.php]: This is kinda confusing so talking through it....

Basically this line is checking the CRM_Core_StateMachine has a urlPath AND the url contains (for drupal) q=blah and the url check is being removed

File: CRM/Core/Controller.php [hunk 3]
```diff
@@ -457,8 +457,8 @@ public function addPages(&$stateMachine, $action = CRM_Core_Action::NONE) {
...
+        $$stateName->urlPath = explode('/', (string) CRM_Utils_System::getUrlPath());
```
Comment (eileenmcnaughton) [inline CRM/Core/Controller.php]: So, here we are setting CRM_Core_StateMachine->urlPath to the 'q' value (for drupal) or NULL if there is not one 

(question - could the property exist and be set & be reset to NULL by this?)

File: CRM/Core/Resources.php [hunk 4]
```diff
@@ -781,11 +781,15 @@ public function coreResourceList($region) {
...
+    if (in_array(CRM_Utils_Array::value('snippet', $_REQUEST), array(
         CRM_Core_Smarty::PRINT_SNIPPET,
         CRM_Core_Smarty::PRINT_NOFORM,
         CRM_Core_Smarty::PRINT_JSON,
-      ));
+      ))
+    ) {
+      return TRUE;
+    }
+    return strpos(CRM_Utils_System::getUrlPath(), 'civicrm/ajax') === 0;
```
Comment (eileenmcnaughton) [inline CRM/Core/Resources.php]: So this seems to be the key change in the PR - ie it can be ajax mode by virtue of having snippet in the url or by the path being civicrm/ajax

Comment (eileenmcnaughton) [inline CRM/Core/Resources.php]: I can definitely confirm there are a lot of ajax paths & checking for example civicrm/ajax/checkphone it is being called from the template without snippet=1 AND it is calling an ajax function - so it should be treated as ajax mode.

I guess there could be an edge case risk that someone put ajax in a url that is not an ajax page. I spot checked a handful & didn't find any

**Comment by eileenmcnaughton** [comment]
I'm confused about those test fails

File: CRM/Core/Controller.php (hunk 2 above)
Comment (colemanw) [inline CRM/Core/Controller.php]: It's just cleanup to use the new utility function.

File: CRM/Core/Controller.php (hunk 3 above)
Comment (colemanw) [inline CRM/Core/Controller.php]: Well this property is something I added pretty recently and it's rarely used. I'm not worried about this cleanup breaking anything.
One of the few places I can think of where this variable is used is places like `/civicrm/admin/options/activity_type` where the form reads the last argument from the url. That's still working fine.

File: CRM/Core/Resources.php (hunk 4 above)
Comment (colemanw) [inline CRM/Core/Resources.php]: Yes this is the meat of it. I think it's safe to assume that any path starting with `/civicrm/ajax` is a callback not a page.

**Comment by colemanw** [comment]
@civicrm-builder retest this please.

**Comment by eileenmcnaughton** [comment]
I've been running this locally & notice no difference 'going about my daily business' & my reading of the PR & the explanations by Coleman suggest it is all good. I'm happy for it to be merged - except that the test issue is still there

**Comment by colemanw** [comment]
@eileenmcnaughton I'm confused too. The tests appear to be passing in the jenkins grid on master... however, when I run them locally on master branch (without these patches) they fail. I can't figure out why this PR would bring that failure to light.

**Comment by eileenmcnaughton** [comment]
OK the failing test interacts with $_REQUEST  - yours runs before it & does this...

$_REQUEST = $_GET = $query;

Perhaps something gets unset. I guess a test would be to return $REQUEST (& maybe GET)  to what it was at the end of the test

**Comment by colemanw** [comment]
Ok, maybe this will fix it...
